taichi/transforms/auto_diff.cpp

```diff
@@ -3,13 +3,185 @@
+      current_ib = block;
+      block->accept(this);
+    } else {
+      // No need to dive further
+    }
+  }
+
+  void visit(StructForStmt *stmt) {
+    TI_ASSERT(depth == 0);
+    depth++;
+    current_ib = stmt->body.get();
+    visit_loop_body(stmt->body.get());
+    depth--;
+    if (depth == 0) {
+      independent_blocks.push_back(current_ib);
+    }
+  }
+
+  void visit(RangeForStmt *stmt) {
+    if (depth == 0) {
+      current_ib = stmt->body.get();
+    }
+    depth++;
+    visit_loop_body(stmt->body.get());
+    depth--;
+    if (depth == 0) {
+      independent_blocks.push_back(current_ib);
+    }
+  }
+
+  static std::vector<Block *> run(IRNode *root) {
+    IdentifyIndependentBlocks pass;
+    Block *block = root->as<Block>();
+    bool has_for = false;
+    for (auto &s : block->statements) {
+      if (s->is<StructForStmt>() || s->is<RangeForStmt>()) {
+        has_for = true;
+      }
+    }
+    if (!has_for) {
+      // The whole block is an IB
+      pass.independent_blocks.push_back(block);
+    } else {
+      root->accept(&pass);
+    }
+    TI_ASSERT(!pass.independent_blocks.empty());
+    return pass.independent_blocks;
+  }
+
+ private:
+  std::vector<Block *> independent_blocks;
+  int depth{0};
+  Block *current_ib{nullptr};
+};
+
+// Note that SSA does not mean the instruction will be executed at most once.
+// For instructions that may be executed multiple times, we treat them as a
+// mutable local variables.
+class PromoteSSA2LocalVar : public BasicStmtVisitor {
+  using BasicStmtVisitor::visit;
+
+  PromoteSSA2LocalVar(Block *block) {
+    alloca_block = block;
+    invoke_default_visitor = true;
+    execute_once = true;
+  }
+
+  void visit(Stmt *stmt) override {
+    if (execute_once)
+      return;
+    TI_ASSERT(stmt->width() == 1);
+    if (!(stmt->is<UnaryOpStmt>() || stmt->is<BinaryOpStmt>() ||
+          stmt->is<TernaryOpStmt>() || stmt->is<OffsetAndExtractBitsStmt>() ||
+          stmt->is<GlobalLoadStmt>())) {
+      // TODO: this list may be incomplete
+      return;
+    }
+    // Create a alloc
+    auto alloc = Stmt::make<AllocaStmt>(1, stmt->ret_type.data_type);
+    auto alloc_ptr = alloc.get();
+    TI_ASSERT(alloca_block);
+    alloca_block->insert(std::move(alloc), 0);
+    auto load = stmt->insert_after_me(
+        Stmt::make<LocalLoadStmt>(LocalAddress(alloc_ptr, 0)));
+    irpass::replace_all_usages_with(stmt->parent, stmt, load);
+    // Create the load first so that the operand of the store won't get replaced
+    stmt->insert_after_me(Stmt::make<LocalStoreStmt>(alloc_ptr, stmt));
+  }
+
+  void visit(RangeForStmt *stmt) override {
+    auto old_execute_once = execute_once;
+    execute_once = false;  // loop body may be executed many times
+    stmt->body->accept(this);
+    execute_once = old_execute_once;
+  }
+
+ private:
+  Block *alloca_block{nullptr};
+  bool execute_once;
+
+ public:
+  static void run(Block *block) {
+    PromoteSSA2LocalVar pass(block);
```

Comment:
`block` won't contain any `StructForStmt`s, right?

Reply:
Of course, `StructFor` can only appear at the top level.